internal/forwarder/gtp5g_test.go

```diff
@@ -278,3 +281,189 @@ func TestGtp5g_CreateRules(t *testing.T) {
+				},
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_DIRECTION,
+					Value: nl.AttrU8(gtp5gnl.SDF_FILTER_OUT),
+				},
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_SRC_IPV4,
+					Value: nl.AttrBytes(net.IPv4(50, 60, 0, 0).To4()),
+				},
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_DEST_IPV4,
+					Value: nl.AttrBytes(net.IPv4(10, 20, 30, 0).To4()),
+				},
+			},
+			err: nil,
+		},
+		{
+			name:       "source port (DL)",
+			s:          "permit out ip from 10.20.30.40/24 345,789-792,1023-1026 to 50.60.70.80/16 456-458,1088,1089",
+			swapSrcDst: false,
+			attrs: nl.AttrList{
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_ACTION,
+					Value: nl.AttrU8(gtp5gnl.SDF_FILTER_PERMIT),
+				},
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_DIRECTION,
+					Value: nl.AttrU8(gtp5gnl.SDF_FILTER_OUT),
+				},
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_SRC_IPV4,
+					Value: nl.AttrBytes(net.IPv4(10, 20, 30, 0).To4()),
+				},
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_DEST_IPV4,
+					Value: nl.AttrBytes(net.IPv4(50, 60, 0, 0).To4()),
+				},
+				nl.Attr{
+					Type: gtp5gnl.FLOW_DESCRIPTION_SRC_PORT,
+					Value: nl.AttrBytes(convertSlice([][]uint16{
+						{345},
+						{789, 792},
+						{1023, 1026},
+					})),
+				},
+				nl.Attr{
+					Type: gtp5gnl.FLOW_DESCRIPTION_DEST_PORT,
+					Value: nl.AttrBytes(convertSlice([][]uint16{
+						{456, 458},
+						{1088},
+						{1089},
+					})),
+				},
+			},
+			err: nil,
+		},
+		{
+			name:       "source port (UL)",
+			s:          "permit out ip from 10.20.30.40/24 345,789-792,1023-1026 to 50.60.70.80/16 456-458,1088,1089",
+			swapSrcDst: true,
+			attrs: nl.AttrList{
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_ACTION,
+					Value: nl.AttrU8(gtp5gnl.SDF_FILTER_PERMIT),
+				},
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_DIRECTION,
+					Value: nl.AttrU8(gtp5gnl.SDF_FILTER_OUT),
+				},
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_SRC_IPV4,
+					Value: nl.AttrBytes(net.IPv4(50, 60, 0, 0).To4()),
+				},
+				nl.Attr{
+					Type:  gtp5gnl.FLOW_DESCRIPTION_DEST_IPV4,
+					Value: nl.AttrBytes(net.IPv4(10, 20, 30, 0).To4()),
+				},
+				nl.Attr{
+					Type: gtp5gnl.FLOW_DESCRIPTION_SRC_PORT,
+					Value: nl.AttrBytes(convertSlice([][]uint16{
+						{456, 458},
+						{1088},
+						{1089},
+					})),
+				},
+				nl.Attr{
+					Type: gtp5gnl.FLOW_DESCRIPTION_DEST_PORT,
+					Value: nl.AttrBytes(convertSlice([][]uint16{
+						{345},
+						{789, 792},
+						{1023, 1026},
+					})),
+				},
+			},
+			err: nil,
+		},
+	}
+
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			attrs, err := g.newFlowDesc(tt.s, tt.swapSrcDst)
+			if tt.err == nil {
+				if err != nil {
+					t.Fatal(err)
+				}
+				assert.Subset(t, attrs, tt.attrs)
+			} else if err != tt.err {
+				t.Errorf("wantErr %v; but got %v", tt.err, err)
+			}
+		})
+	}
+}
+
+// TODO
+// Test on newSdfFilter()
```

Comment:
I leave this to TODO since it's more complicated. 
The others are ready for review. 
@ianchen0119